src/nunavut/lang/cpp/templates/_fields_as_union.j2

```diff
@@ -3,17 +3,77 @@
  # Copyright (C) 2021  UAVCAN Development Team  <uavcan.org>
  # This software is distributed under the terms of the MIT License.
 #}
-    union // The union is placed first to ensure that the active element address equals the object address.
+    class VariantType final
     {
-{%- for field in composite_type.fields_except_padding -%}
-        {{ field.doc | block_comment('cpp-doxygen', 8, 120) }}
-        {{ field.data_type | declaration }} {{ field.name | id }};
+    public:
+        VariantType()
+            : tag_(0)
+            , internal_union_value_()
+        {}
+
+        //
+        // Rule of Five.
+        //
+        VariantType(const VariantType&) = delete;
+        VariantType& operator=(const VariantType&) = delete;
+        VariantType(VariantType&&) = delete;
+        VariantType& operator=(VariantType&&) = delete;
+
+        ~VariantType()
+        {
+            destroy_current();
+        }
+
+        struct IndexOf final
+        {
+            IndexOf() = delete;
+{%- for field in composite_type.fields_except_padding %}
+            static constexpr const std::size_t {{ field.name | id }} = {{ loop.index0 }}U;
 {%- endfor %}
+        };
+
+        template<std::size_t I, class...Types> struct alternative;
+
+{% for field in composite_type.fields_except_padding %}
+        template<class...Types> struct alternative<{{ loop.index0 }}U, Types...>
+        {
+            using type = {{ field.data_type | declaration }};
+        };
+{%- endfor %}
+
+        template< std::size_t I, class... Args > typename VariantType::alternative<I, VariantType>::type& emplace(Args&&... v);
+
+        template<std::size_t I, class... Types>
+        static constexpr typename VariantType::alternative<I, VariantType>::type* get_if(VariantType* v) noexcept;
+
+        template<std::size_t I, class... Types>
+        static constexpr const typename VariantType::alternative<I, VariantType>::type* get_if(const VariantType* v) noexcept;
+
+    private:
+
+        void destroy_current()
+        {
+{%- for field in composite_type.fields_except_padding if field is not PrimitiveType %}
+            {% if not loop.first %}else {% endif %}if (tag_ == {{ loop.index0 }})
+            {
+                if (!std::is_trivially_destructible<{{ field.data_type | declaration }}>::value)
```

Comment:
What is the purpose of this check? If the destructor is indeed trivial, it should be elided by the compiler anyway.

Reply:
Yup. Removed